neural_speed/models/llama/llama.cpp

```diff
@@ -347,17 +356,64 @@ static bool llama_model_eval_internal(model_context* ctx, const model_input* inp
         // cur = cur*ffn_norm(broadcasted)
         cur = ne_mul(ctx0, cur, model.layers[il].norm[1]);
       }
-
-      if (bestla_fusion_FFN_SiLu_f32f32_support(model.layers[il].ffn[0]->data, model.layers[il].ffn[1]->data,
-                                                model.layers[il].ffn[2]->data, N, cur->ne[0],
-                                                model.layers[il].ffn[0]->ne[1], model.layers[il].ffn[1]->ne[1])) {
-        cur = ne_ffn_silu(ctx0, model.layers[il].ffn[0], model.layers[il].ffn[1], model.layers[il].ffn[2], cur);
+      if (n_expert == 0) {
+        if (bestla_fusion_FFN_SiLu_f32f32_support(model.layers[il].ffn[0]->data, model.layers[il].ffn[1]->data,
+                                                  model.layers[il].ffn[2]->data, N, cur->ne[0],
+                                                  model.layers[il].ffn[0]->ne[1], model.layers[il].ffn[1]->ne[1])) {
+          cur = ne_ffn_silu(ctx0, model.layers[il].ffn[0], model.layers[il].ffn[1], model.layers[il].ffn[2], cur);
+        } else {
+          struct ne_tensor* tmp = ne_mul_mat(ctx0, model.layers[il].ffn[2], cur);
+          cur = ne_mul_mat(ctx0, model.layers[il].ffn[0], cur);
+          cur = ne_silu(ctx0, cur);
+          cur = ne_mul(ctx0, cur, tmp);
+          cur = ne_mul_mat(ctx0, model.layers[il].ffn[1], cur);
+        }
       } else {
-        struct ne_tensor* tmp = ne_mul_mat(ctx0, model.layers[il].ffn[2], cur);
-        cur = ne_mul_mat(ctx0, model.layers[il].ffn[0], cur);
-        cur = ne_silu(ctx0, cur);
-        cur = ne_mul(ctx0, cur, tmp);
-        cur = ne_mul_mat(ctx0, model.layers[il].ffn[1], cur);
+        ne_tensor* logits = ne_mul_mat(ctx0, model.layers[il].ffn_gate_inp, cur);  // [n_tokens, num_experts]
+        ne_tensor* probs = ne_soft_max_inplace(ctx0, logits);
+        ne_tensor* selected_experts = ne_top_k(ctx0, probs, n_expert_used);
+        ne_build_forward_expand(&gf, selected_experts);
```

Comment:
Unnecessary to explicitly build graph for `selected_experts` as it is an operand of `ne_get_rows` ?

Reply:
done